### PR TITLE
Knowledge distillation tutorial

### PR DESCRIPTION
**Pull request by lindawangg**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

#### Changelog
What are the changes made in this PR?
* Adds a knowledge distillation tutorial on how to distill Llama3.1 8B into Llama3.2 1B

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)
image

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,228 @@
+.. _llama_kd_label:
+
+============================
+Distilling Llama3 8B into 1B
+============================
+
+This guide will teach you about knowledge distillation (KD) and show you how you can use torchtune to distill a Llama3.1 8B model into Llama3.2 1B.
+If you already know what knowledge distillation is and want to get straight to running your own distillation in torchtune,
+you can jump to knowledge distillation recipe in torchtune, `knowledge_distillation_single_device.py <https://github.com/pytorch/torchtune/blob/main/recipes/knowledge_distillation_single_device.py>`_.
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:9]: nit
```suggestion
you can jump to the knowledge distillation recipe in torchtune, `knowledge_distillation_single_device.py <https://github.com/pytorch/torchtune/blob/main/recipes/knowledge_distillation_single_device.py>`_.
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+How does Knowledge Distillation work?
+-------------------------------------
+
+Knowledge is transferred from the teacher to student model by training it on a transfer set and where the
+student is trained to imitate the token-level probability distributions of the teacher. The diagram below
+is a simplified representation of how KD works.
+
+.. image:: /_static/img/kd-simplified.png
+
+The total loss can be configured in many ways. The default KD config in torchtune combines CE loss with
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:46]:
```suggestion
The total loss can be configured in many ways. The default KD config in torchtune combines the cross-entropy (CE) loss with
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+forward `Kullback-Leibler (KL) divergence <https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence>`_ loss,
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:47]:
```suggestion
the forward `Kullback-Leibler (KL) divergence <https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence>`_ loss,
```

Comment (SalmanMohammadi) [reply docs/source/tutorials/llama_kd_tutorial.rst:47]: nit: could you link to a more concrete reference of this specific loss? e.g. https://arxiv.org/html/2402.03898v2#S2

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+which is used in standard KD approaches.Forward KL divergence aims to minimize the difference by forcing the student
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:48]:
```suggestion
which is used in standard KD approaches. iForward KL divergence aims to minimize the difference by forcing the student
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+      prod_probs = teacher_prob * student_logprob
+      # Compute the sum
+      x = torch.sum(prod_probs, dim=-1).view(-1)
+      # We don't want to include the ignore labels in the average
+      mask = (labels != self.ignore_index).int()
+      # Loss is averaged over non-ignored targets
+      return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)
+
+There are some details omitted to simplify the computation, but if you'd like to know more,
+you can see the implementation in `ForwardKLLoss <https://github.com/pytorch/torchtune/blob/4234b78b914af23384ce0348f564e2119d107a96/torchtune/modules/loss/kd_losses.py>`_.
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:81]: should be able to do
```suggestion
you can see the implementation in :class:`torchtune.modules.loss.ForwardKLLoss`.
```

File: docs/source/tutorials/llama_kd_tutorial.rst (hunk 4 above)
Comment (SalmanMohammadi) [reply docs/source/tutorials/llama_kd_tutorial.rst:48]: which difference are you referring to? : )

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+The current implementation only supports student and teacher models that have the same output
+logit shape and same tokenizer.
+
+KD recipe in torchtune
+----------------------
+
+With torchtune, we can easily apply knowledge distillation to Llama3, as well as other LLM model families.
+Let's take a look at how you could distill a model using torchtune's `KD recipe <https://github.com/pytorch/torchtune/blob/4234b78b914af23384ce0348f564e2119d107a96/recipes/knowledge_distillation_single_device.py>`_.
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:89]: Not sure if you need to link to the recipe again

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+First, make sure that you have downloaded the Llama3 weights. For this example, we'll use the Llama3.1-8B as teacher and Llama3.2-1B as student.
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:91]:
```suggestion
First, make sure that you have downloaded the Llama3.1 weights. For this example, we'll use the Llama3.1-8B as teacher and Llama3.2-1B as student.
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+.. code-block:: bash
+
+    tune download meta-llama/Meta-Llama-3.1-8B-Instruct --output-dir /tmp/Meta-Llama-3.1-8B-Instruct --ignore-patterns "original/consolidated.00.pth" --hf_token <HF_TOKEN>
+
+    tune download meta-llama/Llama-3.2-1B-Instruct --output-dir /tmp/Llama-3.2-1B-Instruct --ignore-patterns "original/consolidated.00.pth" --hf_token <HF_TOKEN>
+
+Then, we will fine-tune the teacher model with using LoRA. Based on our experiments and previous work,
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:99]: nit nit (ignore if you like): ```suggestion
Then, we will fine-tune the teacher model with using LoRA. Based on our experiments and previous work,
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+.. image:: /_static/img/kd-finetune-student.png
+
+Based on the loss graphs, using a fine-tuned teacher model results in a lower loss irrespective of
+whether the student model is fine-tuned or not. It's also interesting to note that the class loss
+starts to increase when using a fine-tuned student model.
+
+Hyperparameter tuning: learning rate
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, the config has the learning rate as 3e-4, same as LoRA configs. For these experiments,
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:189]:
```suggestion
By default, the config has the learning rate as ``3e-4`` which is the same as the LoRA configs. For these experiments,
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,228 @@
+.. _llama_kd_label:
+
+============================
+Distilling Llama3 8B into 1B
+============================
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:5]:
```suggestion
==================================================
Distilling Llama3.1 8B into Llama3.2 1B using Knowledge Distillation
==================================================
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+Based on the loss graphs, using a fine-tuned teacher model results in a lower loss irrespective of
+whether the student model is fine-tuned or not. It's also interesting to note that the class loss
+starts to increase when using a fine-tuned student model.
+
+Hyperparameter tuning: learning rate
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, the config has the learning rate as 3e-4, same as LoRA configs. For these experiments,
+we changed the learning rate from as high as 1e-3 to as low as 1e-5. To change the learning rate,
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:190]: mega nit: should be able to use latexify this using something like :math:`1e^{-1}`

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+Hyperparameter tuning: KD ratio
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the config, we have the kd_ratio as 0.5, which gives even weightings to both the class and KD loss. In these experiments,
+we look at the effects of different KD ratios, where 0 only uses the class loss and 1 only uses the KD loss.
+Similar to changing the learning rate, the KD ratio can be adjusted using:
+
+.. code-block:: bash
+
+    tune run knowledge_distillation_single_device --config llama3_2/knowledge_distillation_single_device kd_ratio=[KD_RATIO]
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:212]: Same comment as above

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+
+
+Overall, the evaluation results are slightly better for higher KD ratios.
+
+.. image:: /_static/img/kd-hyperparam-kd-ratio.png
+
+Qwen2 1.5B to 0.5B
+^^^^^^^^^^^^^^^^^^
+
+The KD recipe can also be applied to different model families as well. Here we look at the effect of KD when the number of
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:222]: mega nit: 
```suggestion
The KD recipe can also be applied to different model families. Here we look at the effect of KD when the number of
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+parameters between the teacher and student models are closer. For this experiment, we used Qwen2 1.5B and Qwen2 0.5B, which can be found in
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:223]: nit:
```suggestion
parameters between the teacher and student models are closer. For this experiment, we used Qwen2 1.5B and Qwen2 0.5B, the configs for which can be found in
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+`qwen2/knowledge_distillation_single_device <https://github.com/pytorch/torchtune/blob/4234b78b914af23384ce0348f564e2119d107a96/recipes/configs/qwen2/knowledge_distillation_single_device.yaml>`_
+config. Here we see that alpaca_cleaned_dataset only improves truthful_qa performance and drops the metrics for the other evaluation tasks.
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:225]:
```suggestion
config. Here we see that training on the alpaca cleaned dataset only improves truthful_qa performance and drops the metrics for the other evaluation tasks.
```

**Comment by SalmanMohammadi** [comment]
Thanks so much for adding this. I've been loosely following along your KD work and it's really impressive.

Lots of small nits but this is an awesome tutorial. I especially love all the empirical results.

File: docs/source/tutorials/llama_kd_tutorial.rst (hunk 1 above)
Comment (SalmanMohammadi) [reply docs/source/tutorials/llama_kd_tutorial.rst:9]: Actually, it might be better to just reference the latter parts of this tutorial for people who want to jump ahead rather than pointing to the recipe file,

File: docs/source/tutorials/llama_kd_tutorial.rst (hunk 5 above)
Comment (SalmanMohammadi) [inline docs/source/tutorials/llama_kd_tutorial.rst:81]: Is it perhaps worth mentioning that we actually use ForwardKLWithChunkedOutputLoss by default?

File: docs/source/tutorials/llama_kd_tutorial.rst (hunk 1 above)
Comment (lindawangg) [reply docs/source/tutorials/llama_kd_tutorial.rst:9]: Changed to link the tutorial section instead of recipe.

File: docs/source/tutorials/llama_kd_tutorial.rst (hunk 11 above)
Comment (lindawangg) [reply docs/source/tutorials/llama_kd_tutorial.rst:190]: Changed all to use math.

File: docs/source/tutorials/llama_kd_tutorial.rst (hunk 5 above)
Comment (lindawangg) [reply docs/source/tutorials/llama_kd_tutorial.rst:81]: added a note that we used the chunked loss to save memory.

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+      * Be familiar with :ref:`torchtune<overview_label>`
+      * Make sure to :ref:`install torchtune<install_label>`
+      * Make sure you have downloaded the :ref:`Llama3 model weights<download_llama_label>`
+      * Be familiar with :ref:`LoRA<lora_finetune_label>`
+
+What is Knowledge Distillation?
+-------------------------------
+
+`Knowledge Distillation <https://arxiv.org/pdf/1503.02531>`_ is is a widely used compression technique
```
Comment (ebsmothers) [inline docs/source/tutorials/llama_kd_tutorial.rst:30]:
```suggestion
`Knowledge Distillation <https://arxiv.org/pdf/1503.02531>`_ is a widely used compression technique
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+that transfers knowledge from a larger (teacher) model to a smaller (student) model. Larger models have
+more parameters and capacity for knowledge, however, this larger capacity is also more computationally
+expensive to deploy. Knowledge distillation can be used to compress the knowledge of a larger model into
+a smaller model. The idea is that performance of smaller models can be improved by learning from larger
+model's outputs.
+
+How does Knowledge Distillation work?
+-------------------------------------
+
+Knowledge is transferred from the teacher to student model by training it on a transfer set and where the
```
Comment (ebsmothers) [inline docs/source/tutorials/llama_kd_tutorial.rst:40]:
```suggestion
Knowledge is transferred from the teacher to student model by training it on a transfer set where the
```

File: docs/source/tutorials/llama_kd_tutorial.rst [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+      # Compute the sum
+      x = torch.sum(prod_probs, dim=-1).view(-1)
+      # We don't want to include the ignore labels in the average
+      mask = (labels != self.ignore_index).int()
+      # Loss is averaged over non-ignored targets
+      return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)
+
+There are some details omitted to simplify the computation, but if you'd like to know more,
+you can see the implementation in :class:`torchtune.modules.loss.ForwardKLLoss`.
+By default, the KD configs use :class:`torchtune.modules.loss.ForwardKLWithChunkedOutputLoss` to reduce memory.
```
Comment (ebsmothers) [inline docs/source/tutorials/llama_kd_tutorial.rst:82]: I think you need to add these APIs to the rst file [here](https://github.com/pytorch/torchtune/blob/4efd7fdb48677a4ac6d78f394be7b4ecfbabf7de/docs/source/api_ref_modules.rst?plain=1#L35). This way it'll render as an actual pointer to our API ref

**Review by SalmanMohammadi — APPROVED** [review]

File: docs/source/tutorials/llama_kd_tutorial.rst (hunk 18 above)
Comment (lindawangg) [reply docs/source/tutorials/llama_kd_tutorial.rst:82]: added. Thanks for the pointer.

**Comment by SalmanMohammadi** [comment]
LGTM - please acquire Evan's blessing before merging.

**Review by ebsmothers — APPROVED** [review]
🙌
